packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts

```diff
@@ -433,6 +467,167 @@ export class GraphQLApi extends Construct {
+      grantee,
+      actions,
+      resourceArns: resources.map((resource) => {
+        return this.generateResourceArn(resource);
+      }),
+      scope: this,
+    });
+
+  }
+
+  /**
+   * Adds an IAM policy statement for full access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal
+   */
+  public grantFullAccess(grantee: IGrantable): Grant {
+    return Grant.addToPrincipal({
+      grantee,
+      actions: ['appsync:GraphQL'],
+      resourceArns: [
+        this.generateResourceArn({custom: '*'}),
+      ],
+      scope: this,
+    });
+  }
+
+  /**
+   * Adds an IAM policy statement for Mutation access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal
+   * @param fields The fields to grant access to that are Mutations (leave blank for all)
+   */
+  public grantMutation(grantee: IGrantable, fields?: string[]): Grant {
+    return this.grantType(grantee, 'Mutation', fields);
+  }
+
+  /**
+   * Adds an IAM policy statement for Query access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal
+   * @param fields The fields to grant access to that are Queries (leave blank for all)
+   */
+  public grantQuery(grantee: IGrantable, fields?: string[]): Grant {
+    return this.grantType(grantee, 'Query', fields);
+  }
+
+  /**
+   * Adds an IAM policy statement for Subscription access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal
+   * @param fields The fields to grant access to that are Subscriptions (leave blank for all)
+   */
+  public grantSubscription(grantee: IGrantable, fields?: string[]): Grant {
+    return this.grantType(grantee, 'Subscription', fields);
+  }
+
+  /**
+   * Adds an IAM policy statement for Type access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal
+   * @param type The type to grant access
+   * @param fields The fields to grant access to that are of @param type (leave blank for all)
+   */
+  public grantType(grantee: IGrantable, type: string, fields?: string[]): Grant {
+    return Grant.addToPrincipal({
+      grantee,
+      actions: ['appsync:GraphQL'],
+      resourceArns: fields ?
+        fields.map((field) => {
+          return this.generateResourceArn({ type: type, field: field });
+        }) :
+        [
+          this.generateResourceArn({ type: type }),
+        ],
+      scope: this,
+    });
+  }
+
+  /**
+   * Generate a Resource ARN to use for grant functions.
+   *
+   * If custom is specified, then resourceName is custom.
+   * If type is empty, generate full access ('*').
+   * If field is empty, generate full access to specified type.
+   *
+   * @param resource - an singular IAM resource to provision
+   */
+  private generateResourceArn(resource: IamResources): string {
+    const name = resource.custom ? resource.custom :
```

Comment:
So the logic is:

1. If `custom` is set, all other properties are ignored.
2. If `type` is set, you can also set `field`.
3. If no property is set, it's equivalent to `'*'`.

I actually don't think an interface is a good match for this then. How about something like this instead?

```ts
export class IamResources {
  public static custom(custom: string): IamResources { ... }

  public static ofType(type: string, field?: string): IamResources { ... }

  public static all(): IamResources { ... } // equivalent to '*'


  private constructor IamResources() { ... }

  public name(): string {
    // move the logic to this class
  }
}
```

?

Reply:
yeah thats the right logic!

Would this replace the entire `grant` logic? I think the `grant` functions fall more similar to the rest of CDK's permissions stuff. As opposed to an interface or class.. but I'm assuming this is more towards the logic of creating the `resourceArn` so something like

```ts
api.grant(role, new IamResource.ofType('Mutation', 'addTest')); 
```

could be called?

Reply:
Yes .Except no `new` (`ofType` is  static method).

Reply:
merging this comment with the later one.. I think you're right this is definitely easier to maneuver. Out of curiosity, when is it good to do an interface vs an object?

Reply:
It's hard to say precisely, but a good guideline might be: if you need some behavior, use a class. If you just need a bag of properties, use an interface.

In this case, the different properties of the interface had very strong dependencies between each other (and certain invariants), so a class definitely seems like the way to go.

Reply:
Also, if an interface without any properties has a significant meaning (like here), you definitely want a class (otherwise you're forcing users to provide you `{}`, which is super awkward).